rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/AbstractLMDBStore.scala

```diff
@@ -0,0 +1,64 @@
+package coop.rchain.rspace.nextgenrspace.history
+
+import java.nio.ByteBuffer
+
+import cats.effect.Sync
+import org.lmdbjava.{Dbi, Env, Txn}
+
+import scala.util.control.NonFatal
+
+abstract class AbstractLMDBStore[F[_]: Sync] {
+  def env: Env[ByteBuffer]
+  def dbi: Dbi[ByteBuffer]
+
+  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+  // TODO stop throwing exceptions
+  private def withTxn[R](txn: Txn[ByteBuffer])(f: Txn[ByteBuffer] => R): R =
+    try {
+      val ret: R = f(txn)
+      txn.commit()
+      ret
+    } catch {
+      case NonFatal(ex) =>
+        ex.printStackTrace()
+        throw ex
+    } finally {
+      txn.close()
+    }
+
+  private def createTxnRead() =
+    env.txnRead
+
+  private def createTxnWrite() =
+    env.txnWrite
+
+  private def withReadTxnF[R](f: Txn[ByteBuffer] => R): F[R] =
+    Sync[F].delay {
+      withTxn(createTxnRead())(f)
+    }
+
+  def withWriteTxnF[R](f: Txn[ByteBuffer] => R): F[R] =
+    Sync[F].delay {
+      withTxn(createTxnWrite())(f)
+    }
+
+  def get(key: ByteBuffer): F[ByteBuffer] =
+    withReadTxnF { txn =>
+      dbi.get(txn, key)
+    }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+  def put(key: ByteBuffer, value: ByteBuffer): F[Unit] =
+    withWriteTxnF { txn =>
+      if (dbi.put(txn, key, value)) {
+        ()
+      } else {
+        throw new RuntimeException("was not able to put data")
```

Comment:
raiseError will not invoke the "finally" block.

Reply:
well, if we move away from exceptions we should use cats Resource instead of try/catch.

Now your response makes me wonder: if we use try/catch/finally mixed with F[_], are the runtime errors properly handled?